sdk/ai/azure-ai-inference/src/main/java/com/azure/ai/inference/ChatCompletionClientTracer.java

```diff
@@ -0,0 +1,606 @@
+
+        final Mono<StreamingChatCompletionsState> resourceSupplier = Mono.fromSupplier(() -> {
+            final StreamingChatCompletionsState resource = state;
+
+            final Context span
+                = tracer.start(rootSpanName(resource.request), new StartSpanOptions(SpanKind.CLIENT), Context.NONE);
+            traceCompletionRequestAttributes(resource.request, span);
+            traceCompletionRequestEvents(resource.request.getMessages(), span);
+            return resource.setSpan(span);
+        });
+
+        final Function<StreamingChatCompletionsState, Flux<StreamingChatCompletionsUpdate>> resourceClosure
+            = resource -> {
+                final RequestOptions rOptions = resource.requestOptions.setContext(resource.span);
+                final Flux<StreamingChatCompletionsUpdate> completionChunks
+                    = resource.operation.invoke(resource.completeRequest, rOptions);
+                return completionChunks.doOnNext(resource::onNextChunk);
+            };
+
+        final Function<StreamingChatCompletionsState, Mono<Void>> asyncComplete = resource -> {
+            final Context span = resource.span;
+            final StreamingChatCompletionsUpdate lastChunk = resource.lastChunk;
+            final String finishReasons = resource.getFinishReasons();
+
+            traceCompletionResponseAttributes(lastChunk, finishReasons, span);
+            traceChoiceEvent(resource.toJson(logger), OffsetDateTime.now(ZoneOffset.UTC), span);
+            tracer.end(null, null, span);
+            return Mono.empty();
+        };
+
+        final BiFunction<StreamingChatCompletionsState, Throwable, Mono<Void>> asyncError = (resource, throwable) -> {
+            final Context span = resource.span;
+
+            tracer.setAttribute("error.type", throwable.getClass().getName(), span);
+            traceChoiceEvent(FINISH_REASON_ERROR, OffsetDateTime.now(ZoneOffset.UTC), span);
+            tracer.end(null, throwable, span);
+            return Mono.empty();
+        };
+
+        final Function<StreamingChatCompletionsState, Mono<Void>> asyncCancel = resource -> {
+            final Context span = resource.span;
+            tracer.setAttribute("error.type", "cancelled", span);
+            traceChoiceEvent(FINISH_REASON_CANCELED, OffsetDateTime.now(ZoneOffset.UTC), span);
+            tracer.end("cancelled", null, span);
+            return Mono.empty();
+        };
+
+        return Flux.usingWhen(resourceSupplier, resourceClosure, asyncComplete, asyncError, asyncCancel);
+    }
+
+    //<editor-fold desc="Private util types, methods">
+    private String rootSpanName(ChatCompletionsOptions completeRequest) {
+        return CoreUtils.isNullOrEmpty(completeRequest.getModel()) ? "chat" : "chat " + completeRequest.getModel();
+    }
+
+    private void traceCompletionRequestAttributes(ChatCompletionsOptions request, Context span) {
+        final String modelId = request.getModel();
```

Comment:
we can check if span is recording with https://github.com/Azure/azure-sdk-for-java/blob/b8a9d8c1da8114ed9ed1c451a0549f56d4d9b510/sdk/core/azure-core/src/main/java/com/azure/core/util/tracing/Tracer.java#L650

and if not, avoid adding attributes and events to it.

This is a perf optimization that works with sampling. I.e. the tracer is enabled, but only e.g. 1% of spans is sampled in (and will be recorded), so it does not make sense to collect all the data if we know that span won't be recorded.